Add automatic token refresh based on token expiry time

akincel opened the pull request:
Breaking changes:
- `stayLoggedIn()` method is no longer available (background refresh is enabled by default)
- `ensureLoggedIn({ enableLockWidget: bool, forceRefresh: bool })` can be called repeatedly. It will or will not (if there is already valid token) get a new token

Details:
- clients continue retrieving token via `tokenRefreshed` hook but there is also an option of `getLastAuthResult()` which can be used as a token provider. It does not guarantee token availability though
- token together with its expiry time is stored in memory
- scheduled refresh is automatically removed on logout or when login is removed programatically

akincel committed:
Add automatic token refresh based on token expiry time

- `stayLoggedIn()` method is no longer available
  (background refresh is enabled by default)
- token expiry date time is stored in `localStorage`
- scheduled refresh is automatically removed on logout
  or when login is removed programatically

akincel commented:
closes #7

akincel committed:
Use more specific localStorage key for token expiry date

wparad commented on src/auth0-sso-login.js:
```diff
@@ -31,8 +31,10 @@ export class localStorageInteraction {
   }
 }
 
+export const tokenExpiresAtKey = 'cimpress.auth0.tokenExpiresAt';
```
Should we call this to match the library? `cimpress.auth0-sso-login.tokenExpiresAt`

Another question why export this via a function as in `getTokenExpiry`, shouldn't we hide the usage of the `localStorage` to prevent the exposing the implementation details and allowing clients to alter this value directly.

akincel replied:
For other reviewers: Not relevant anymore, use of `localStorage` was dropped.

wparad commented on src/auth0-sso-login.js:
```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
 }
 
+// static localStorage interaction, mostly wrapped for enabling easier mocking through unit tests
+export class localStorageInteraction {
```
I suggest pulling this out into its own class, we could `import` it here and then `mock` it, rather than `stubbing`.  It would provide us assurances the methods don't change.  It also prevents accidental usage of the `localStorage` in the unit tests if we forget to stub.

Since we never use a `key` other than the `tokenExpiresAtKey` I'm almost thinking that should be injected into that class and hide the ability for the library to use different keys.

Something like:
```javascript
const tokenExpiresAtKey = 'cimpress.auth0.tokenExpiresAt';
export default class localStorageIntegration {
  getTokenExpiry() {
    let tokenExpiresAtString = localStorage.getItem(tokenExpiresAtKey);
     return tokenExpiresAtString ? JSON.parse(tokenExpiresAtString) : null;
  },
  setTokenExpiry(value) {
    localStorage.setItem(tokenExpiresAtKey, JSON.stringify(value));
  }
  removeTokenExpiry() {
    localStorage.removeItem(tokenExpiresAtKey);
  }
}
```

wparad replied:
New question, why even use the `localStorage` here.  Is there a scenario where we would want to remember the last refresh after the site was closed?  Doesn't auth have to happen again anyway?  If that is the case, in memory is a better solution.  You get the free `removeItem` and don't have to worry about the store.

akincel replied:
I agree with the first suggestion.

There is a scenario. We have to consider that the library is not triggering initial login on its own. It is triggered from the client (using `ensureLoggedIn`). The client logic checks the validity of the token outside of this library and therefore client skips calling `ensureLoggedIn` on page load. We need a way to schedule the renewal (and check the token expiry). Step by step:
1. Application gets token using `ensureLoggedIn`
2. Library schedules renewal in `tokenRefreshed`
3. User does page refresh (library looses scheduled renewal)
4. Library schedules renewal in `constructor`
5. Application will not call `ensureLoggedIn` as it has valid token from the previous login (no call to `tokenRefreshed`). Therefore, the 4th step was required.

wparad replied:
If we can't guarantee that `ensureLoggedIn` is called, then we can't guarantee that `constructor` is called.  Why not always require `ensureLoggedIn` is called to get all the functionality.  Should we investigate the case where `ensureLoggedIn` isn't not supposed be called.  If we could somehow still assume it would be (possible parameter, or other check) then that would reduce the complexity.

akincel replied:
Actually it is not `ensureLoggedIn` directly, it's about `tokenRefreshed` which is scheduling the renewal after getting new token.

Two questions:
1. Is it feasible to get a new token on every application load? I don't think so.
2. How can we guarantee that token renewal will be registered after application load? I thought `constructor` is the best place we can have. Maybe more aggressively, we can schedule token renewal on `window.event('load')`

akincel replied:
I think we can easily adjust the renewal scheduling to allow calling `ensureLoggedIn`. What really prevents us from doing so is that it will try to get a new token.

The library doesn’t have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

wparad replied:
My reason for this inquiry: constructors with side effects tell me something is off.
To answer your questions:

> 1. Is it feasible to get a new token on every application load? I don't think so.

What code will be executed in this process?  I.e. does application load mean, new tab, F5, hard refresh etc...?  In that case, I would say we actually need to call `ensureLoggedIn` because even if the token is valid, the SSO session may not be.  The token should almost be considered a session specific object, and not saved in a clients application.  It may provide some improvements, such as not showing the `login` button temporarily, but the source of truth is the Auth0 SSO session.

> 2. How can we guarantee that token renewal will be registered after application load? I thought constructor is the best place we can have.

I wouldn't worry about this.  We can always say, to use this library execute, is there a reason we need to be as aggressive?  Is there a reason we need to avoid `ensureLoggedIn`
```javascript
import auth0Manager from 'auth0-sso-login';
auth0Manager.ensureLoggedIn({ avoidAuth0LockWidget: true });
```

Can we follow that line of questioning?  Why does the library need to:
> have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

akincel replied:
If we require `ensureLoggedIn` after importing this library, we increase complexity on the client side. What if their application load does not require auth0 token immediately, but only upon entering a specific protected route. If clients protect the route with `ensureLoggedIn`, they will get a new token on every route entry.

>have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

Knowing the token would solve this. Just let clients call `ensureLoggedIn` whenever they need to know that a valid token is present. This library can decide, whether to get a new one or not.

wparad replied:
(Is one more line of code extra complexity? I think the trade off to not having unintended side-effects on an import statement or constructor is worth the trade off)

>Knowing the token would solve this. Just let clients call ensureLoggedIn whenever they need to know that a valid token is present. This library can decide, whether to get a new one or not.

I agree with this, so figuring out how to get that in is important, but I'm not sure that answers the question of why we need to use the `localStorage`.

I see what the problem is now.  I would expect that `ensureLoggedIn` to be called every time the application is opened, but not always force the user to the auth0Lock.  in other words, we should always try to get a token if possible, but if login isn't required then don't show the lock.

For example take [this google site](https://support.google.com/adwords/?hl=en#topic=3119071) for instance, while login isn't required, you can see from the screen shot that I'm still recognized.  I was expecting this library to perform the same action.

![image](https://user-images.githubusercontent.com/5056218/34034171-a5e8665e-e17d-11e7-83c4-b4094423e4dc.png)


I.e. try to get the token always, and then let the application decide how to handle missing a token.  In other words there are there things that need to be done:
* Get a token if available => Should always happen
* Schedule the refresh => Should always happen
* Get the Auth0 Popup => Should only happen when navigating to a page that requires a valid token.

With the new implementation we are moving the first two into the third one.  But I posit that these are actually separate actions.

If we call `ensureLoggedIn` to be synonymous with the `Get the Auth0 Popup if necessary` that explains why we want to move to the constructor the scheduling of the refresh.

Instead I would suggest:
* On application start call the following code:
```javascript
  import auth0Manager from 'auth0-sso-login';
  const auth0Manager = new Auth0Manager();
  auth0Manager.ensureLoggedIn();
```
* Then on each route change that requires a token
```javascript
   auth0Manager.ensureLoggedIn({ enableLockWidget: true });
```
We could also go the direction towards pulling the logic into the constructor as this PR is taking it, but we'll should still handle getting a new token at the same time as the refresh if possible without triggering the lock widget.

akincel replied:
I see it shaping :) What you say with the parameter to `ensureLoggedIn` makes sense. We can get rid of the `localStorage` as I agree that new application load should get a new token as well. So storing the expiry date in memory is sufficient.

There is still one deal breaker though. Calling ` auth0Manager.ensureLoggedIn({ enableLockWidget: true })` will always get a new token on route change. How to prevent this? It again goes to the direction if this library should store the token on its own or not.

Edit: Instead of token expiry, we can store the whole token in memory.

wparad commented on src/auth0-sso-login.js:
```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
+  if (!tokenExpiresAtJson) {
+    return null;
+  }
+  const tokenExpiresAt = JSON.parse(tokenExpiresAtJson);
```
I'd move this to the `localStoreInteraction` let that instead be the `tokenExpiryManager`.  Either way the variable here probably should be instead `tokenExpiresAtString`

wparad commented on src/auth0-sso-login.js:
```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
   static updateWindow(url) {
```
I'd recommend the one class per file.

akincel replied:
moved to separate file

wparad commented on src/auth0-sso-login.js:
```diff
@@ -155,9 +198,6 @@ export default class auth {
       webAuth.renewAuth(renewOptions, (err, authResult) => {
         if (err) {
           this.log(`Failed to update ID token on retry ${retries}: ${JSON.stringify(err)}`);
-          if (err.error === 'login_required') {
```
If the SSO session in Auth0 has been removed, it means that they probably have logged out of either client of this library or by calling the `https://auth0.com/logout` url, shouldn't we force the user to login again?

akincel replied:
We still do. I moved it out of here, as this is one specific case and we were most likely missing some other (and not removing login).

Just follow the calls in `ensureLoggedIn` and you will find it. The sequence there is easy:
1. Try silent `renewAuth` -> failure
2. Call `removeLogin`, so that client has a chance to react to auth0 lock open and no valid token available (e.g. stop scheduled calls which use token)
3. Open auth0 lock and force login

wparad replied:
I see, missed that.

wparad commented on src/auth0-sso-login.js:
```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
+
+  // refresh token in the second third of its lifetime
+  const refreshDelay = Math.floor((tokenExpiresAt - Date.now()) / 3) * 2;
+  return windowInteraction.setTimeout(renewFunction, refreshDelay);
```
With `window.setTimeout` If the refresh fails `renewFunction` will never get called a second time.  Rather than rescheduling the refresh as part of the `renumFunction` I suggest using `setInterval` here, and removing the line
`this.tokenRefreshHandle = scheduleRenewal(this.tokenRefreshHandle, () => this.ensureLoggedIn());` from the `tokenRefresh` call.

akincel replied:
If it fails, the resulting state should be auth0 lock open waiting for user interaction. There is no reason to have the refresh interval still running.

If we find out that `setInterval` will be more reliable, we should make sure that it will not open multiple auth0 locks, as it fires while user is entering his credentials.

akincel committed:
Allow repeated calls to `ensureLoggedIn` without getting a new token

The library stores `authResult` in memory, that enables it to:
- return latest `authResult` upon request
- schedule auth refresh based on the token expiration time
- allow `ensureLoggedIn` to be called over and over without retrieving
  a new token

akincel committed:
Improve `ensureLoggedIn` documentation

akincel committed:
Create TokenExpiryManager to handle refresh and remaining time to token expiry

akincel committed:
Update README with the new token retrieval option

akincel committed:
Test different paths in `ensureLoggedIn`

akincel committed:
Test token expiry manager

akincel commented:
@wparad and others feel free to re-review. Created `tokenExpiryManager` and added a bunch of tests.

akincel committed:
Extract auth0Lock factory to its own file

wparad commented on src/auth0-sso-login.js:
```diff
@@ -1,27 +1,36 @@
-    window.location = url;
-  }
-}
+import windowInteraction from './window-interaction';
```
`windowInteraction` isn't used in here right?

akincel replied:
It still is used in `logout` for redirect (`updateWindow`) to auth0 logout url.

wparad commented on src/auth0-sso-login.js:
```diff
@@ -38,32 +47,64 @@ export default class auth {
+    this.authResult = null;
+
     if (this.config) {
       if (this.config.hooks && this.config.hooks.logout) {
```
Should we call the `removeLogin` hooks too?  We might want to look at a client implementation to see if it makes sense to call that.

akincel replied:
For a client, these actions probably mean the same, as the result is the same => no access token available. I can imagine that some clients can have some additional logic for `logout` compared to `removeLogin`. Therefore `removeLogin` calling `logout` hooks could be an issue.

You are proposing the opposite though (`logout` triggering `removeLogin` hooks), which I think won't hurt and it also makes sense that logout removes login :)

wparad replied:
We are actually doing that in our application i.e. calling `removeLogin` hooks on `logout`.

akincel replied:
Done. Logout triggers `removeLogin` hooks as well.

wparad reviewed (commented):
Everything looks mostly good, there is just two small things which could be changed.

akincel committed:
Logout triggers removeLogin hooks as well
